### DISCUSSION TITLE
Better example tooling

### DISCUSSION THREAD

**Discussion by Arithmomaniac** [microsoft/typespec#4517] [Ideas]
Many API developer portals (such as [readme.com](https://readme.com/)) make heavy use of examples, and it's a huge boon to have compile-type checking on them. But writing them is a bit hard.

I can think of two main places examples can come from:
- **Payload first**: you already have some JSON (for example, if you're documenting an existing API) and you want to make sure your TypeSpec definition conforms.
In this case, the migration from e.g. `{"my array": []}` to ``#{`my array`: #[]}`` makes this _super_ tedious in bulk (see #3931). It would be a huge help if there was a CLI command that could take arbitrary JSON and spit out equivalent "value"s.
- **Model first**: you have a TypeSpec model and are trying to write a definition for it. Unfortunately, there is no templating of the `example` decorator, so while there is realtime method validation, intellisense gives you no help (asides from general string matching) in defining the example:
<img src="https://github.com/user-attachments/assets/28b5829a-7e82-4e20-a5ce-44df2f7f7a13" width="200"><img src="https://github.com/user-attachments/assets/3430ac48-39e2-4b58-872d-f0311f25ccd2" width="200">
It would be great if the type of `@example : example` could be an inferred `T` instead of just `unknown`.

And even after you have examples, the differing tooltip and error messages between different ways of specifying an example is really confusing and sometimes incredibly verbose [[Playground]](https://typespec.io/playground?c=Y29uc3QgZm9vZ29vZDEgOiBGb28gPSAje0JhcjrFB3o6ICLFHSJ9fTsNCs0xMiA6IMQqxjhCYXJ91jwy0zwz1iszzytiYfUAl2TFMcQd0jD%2FAJbJO%2FAAlmJh7wCVySrnAJUNCkBleGFtcGxlKOgBMinSFDLTFDPMFPMA9jTEYc845ADd0hPQXuQAntZd6QC%2ByF1tb2RlbOUBSnsNCiAg7AEiOw0KfcshxBvHJHogOiBzdHJpbmfEJA%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%7D). So much so, I'm not sure I have a concrete "ask" (except #4544 and #4545).

_These thoughts are based on the work migrating the Sefaria spec; they have hundreds of examples which (see #3866) were manually migrated._

**Comment by timotheeguerin** [comment]
we talked about that when doing example, we do have a gap in the language to be able to represent that, decided to hold on it for now but I do agree this would be really nice.

**Reply by Arithmomaniac** [reply]
Which "that" are you referring to?

**Reply by timotheeguerin** [reply]
sorry referring the the completion.

For the other part I think what we might be able to do is
1. If its just simple primitive types and hte only problem is the missing `#` we could maybe make the code fix act recursively so you
2. If its about converting a string date to `utcDateTime.fromIso` this might be a bit more complex and not sure how we can provide a generic solution to that.

**Reply by Arithmomaniac** [reply]
It's not just the missing `#`: you have to convert the string keys too.

The code fix turns 
```
@example({"Baz": "Hello"})
model FooBar { Baz : string; }
```

into `#{"Baz": "Hello"}`, but that's not good enough:
> Property 'Baz' is missing on type '{ \<missing identifier\>2: "Baz", \<missing identifier\>3: "Hello" }' but required in 'FooBar'
> TypeSpec(missing-property)

It needs to be recognized as JSON and turned into `#{Baz: "Hello"}`.

I wouldn't let dealing with (2) prevent a fix to that, or (1).

**Reply by Arithmomaniac** [reply]
@timotheeguerin would you like me to refile that last comment as a feature request?

**Reply by timotheeguerin** [reply]
yeah

**Reply by Arithmomaniac** [reply]
Alright, see #4612 and #4613

**Comment by Arithmomaniac** [comment]
At this point, about everything actionable has an issue created for it.
